File: LMGeocoder/LMGeocoder.m
//
//  LMGeocoder.m
//  LMGeocoder
//
//  Created by LMinh on 31/05/2014.
//  Copyright (c) 2014 LMinh. All rights reserved.
//

#import "LMGeocoder.h"
#import "LMAddress.h"

static NSString * const kLMGeocoderErrorDomain = @"LMGeocoderError";

#define kGoogleAPIReverseGeocodingURL(lat, lng) [NSString stringWithFormat:@"https://maps.googleapis.com/maps/api/geocode/json?latlng=%f,%f&sensor=true", lat, lng];
#define kGoogleAPIGeocodingURL(address)         [NSString stringWithFormat:@"https://maps.googleapis.com/maps/api/geocode/json?address=%@&sensor=true", address];
#define kGoogleAPIURLWithKey(url, key)          [NSString stringWithFormat:@"%@&key=%@", url, key];

@interface LMGeocoder ()

@property (nonatomic, strong) CLGeocoder *appleGeocoder;
@property (nonatomic, strong) NSURLSessionDataTask *googleGeocoderTask;

@end

@implementation LMGeocoder

@synthesize isGeocoding = _isGeocoding;

#pragma mark - INIT

+ (LMGeocoder *)sharedInstance
{
    static LMGeocoder *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[LMGeocoder alloc] init];
    });
    return sharedInstance;
}

+ (LMGeocoder *)geocoder
{
    return [[LMGeocoder alloc] init];
}

- (id)init
{
    self = [super init];
    if (self != nil) {
        self.appleGeocoder = [[CLGeocoder alloc] init];
    }
    return self;
}


#pragma mark - GEOCODE

- (void)geocodeAddressString:(NSString *)addressString
                     service:(LMGeocoderService)service
           completionHandler:(LMGeocodeCallback)handler
{
    _isGeocoding = YES;
    
    // Check address string
    if (addressString == nil || addressString.length == 0)
    {
        // Invalid address string --> Return error
        NSError *error = [NSError errorWithDomain:kLMGeocoderErrorDomain
                                             code:kLMGeocoderErrorInvalidAddressString
                                         userInfo:nil];
        
        _isGeocoding = NO;
        if (handler) {
            handler(nil, error);
        }
    }
    else
    {
        // Valid address string --> Check service
        switch (service)
        {
            case kLMGeocoderGoogleService:
            {
                // Geocode using Google service
                NSString *urlString = kGoogleAPIGeocodingURL(addressString);
                if (self.googleAPIKey != nil) {
                    urlString = kGoogleAPIURLWithKey(urlString, self.googleAPIKey)
                }
                [self buildAsynchronousRequestFromURLString:urlString
                                          completionHandler:^(NSArray<LMAddress *> * _Nullable results, NSError * _Nullable error) {
                                              
                                              _isGeocoding = NO;
                                              if (handler) {
                                                  handler(results, error);
                                              }
                                          }];
                break;
            }
            case kLMGeocoderAppleService:
            {
                // Geocode using Apple service
                [self.appleGeocoder geocodeAddressString:addressString
                                       completionHandler:^(NSArray *placemarks, NSError *error) {
                                           
                                           _isGeocoding = NO;
                                           
                                           if (!error && placemarks.count) {
                                               // Request successful --> Parse response results
                                               [self parseGeocodingResponseResults:placemarks service:kLMGeocoderAppleService];
                                           }
                                           else {
                                               // Request failed --> Return error
                                               if (handler) {
                                                   handler(nil, error);
                                               }
                                           }
                                       }];
                break;
            }
            default:
                break;
        }
    }
}

- (nullable NSArray *)geocodeAddressString:(nonnull NSString *)addressString
                                   service:(LMGeocoderService)service
                                     error:(NSError **)error
{
    // Check address string
    if (addressString == nil || addressString.length == 0)
    {
        // Invalid address string --> Return
        *error = [NSError errorWithDomain:kLMGeocoderErrorDomain
                                     code:kLMGeocoderErrorInvalidAddressString
                                 userInfo:nil];
        return nil;
    }
    else
    {
        // Valid address string --> Geocode using Google service
        NSString *urlString = kGoogleAPIGeocodingURL(addressString);
        if (self.googleAPIKey != nil) {
            urlString = kGoogleAPIURLWithKey(urlString, self.googleAPIKey)
        }
        NSArray *finalResults = [self buildSynchronousRequestFromURLString:urlString];
        return finalResults;
    }
}


#pragma mark - REVERSE GEOCODE

- (void)reverseGeocodeCoordinate:(CLLocationCoordinate2D)coordinate
                         service:(LMGeocoderService)service
               completionHandler:(LMGeocodeCallback)handler
{
    _isGeocoding = YES;
    
    // Check location coordinate
    if (!CLLocationCoordinate2DIsValid(coordinate))
    {
        // Invalid location coordinate --> Return error
        NSError *error = [NSError errorWithDomain:kLMGeocoderErrorDomain
                                             code:kLMGeocoderErrorInvalidCoordinate
                                         userInfo:nil];
        
        _isGeocoding = NO;
        if (handler) {
            handler(nil, error);
        }
    }
    else
    {
        // Valid location coordinate --> Check service
        switch (service)
        {
            case kLMGeocoderGoogleService:
            {
                // Reverse geocode using Google service
                NSString *urlString = kGoogleAPIReverseGeocodingURL(coordinate.latitude, coordinate.longitude);
                if (self.googleAPIKey != nil) {
                    urlString = kGoogleAPIURLWithKey(urlString, self.googleAPIKey)
                }
                [self buildAsynchronousRequestFromURLString:urlString
                                          completionHandler:^(NSArray<LMAddress *> * _Nullable results, NSError * _Nullable error) {
                                              
                                              _isGeocoding = NO;
                                              if (handler) {
                                                  handler(results, error);
                                              }
                                          }];
                break;
            }
            case kLMGeocoderAppleService:
            {
                // Reverse geocode using Apple service
                CLLocation *location = [[CLLocation alloc] initWithLatitude:coordinate.latitude
                                                                  longitude:coordinate.longitude];
                [self.appleGeocoder reverseGeocodeLocation:location
                                         completionHandler:^(NSArray *placemarks, NSError *error) {
                                             
                                             _isGeocoding = NO;
                                             
                                             if (!error && placemarks.count) {
                                                 // Request successful --> Parse response results
                                                 [self parseGeocodingResponseResults:placemarks service:kLMGeocoderAppleService];
                                             }
                                             else {
                                                 // Request failed --> Return error
                                                 if (handler) {
                                                     handler(nil, error);
                                                 }
                                             }
                                         }];
                break;
            }
            default:
                break;
        }
    }
}

- (nullable NSArray *)reverseGeocodeCoordinate:(CLLocationCoordinate2D)coordinate
                                       service:(LMGeocoderService)service
                                         error:(NSError **)error
{
    // Check location coordinate
    if (!CLLocationCoordinate2DIsValid(coordinate))
    {
        // Invalid location coordinate --> Return
        *error = [NSError errorWithDomain:kLMGeocoderErrorDomain
                                     code:kLMGeocoderErrorInvalidCoordinate
                                 userInfo:nil];
        return nil;
    }
    else
    {
        // Valid location coordinate --> Reverse geocode using Google service
        NSString *urlString = kGoogleAPIReverseGeocodingURL(coordinate.latitude, coordinate.longitude);
        if (self.googleAPIKey != nil) {
            urlString = kGoogleAPIURLWithKey(urlString, self.googleAPIKey)
        }
        NSArray *finalResults = [self buildSynchronousRequestFromURLString:urlString];
        return finalResults;
    }
}


#pragma mark - CANCEL

- (void)cancelGeocode
{
    if (self.appleGeocoder) {
        [self.appleGeocoder cancelGeocode];
    }
    
    if (self.googleGeocoderTask) {
        [self.googleGeocoderTask cancel];
    }
}


#pragma mark - CONNECTION STUFF

- (void)buildAsynchronousRequestFromURLString:(NSString *)urlString
                            completionHandler:(LMGeocodeCallback)handler
{
    urlString = [urlString stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:[NSURL URLWithString:urlString]];
    
    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:configuration];
    self.googleGeocoderTask = [session dataTaskWithRequest:request
                                         completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                             
                                             if (!error && data)
                                             {
                                                 // Request successful --> Parse response to JSON
                                                 NSError *parsingError = nil;
                                                 NSDictionary *result = [NSJSONSerialization JSONObjectWithData:data
                                                                                                        options:NSJSONReadingAllowFragments
                                                                                                          error:&parsingError];
                                                 if (!parsingError && result)
                                                 {
                                                     // Parse successful --> Check status value
                                                     NSString *status = [result valueForKey:@"status"];
                                                     if ([status isEqualToString:@"OK"])
                                                     {
                                                         // Status OK --> Parse response results
                                                         NSArray *locationDicts = [result objectForKey:@"results"];
                                                         NSArray *finalResults = [self parseGeocodingResponseResults:locationDicts service:kLMGeocoderGoogleService];
                                                         
                                                         if (handler) {
                                                             handler(finalResults, nil);
                                                         }
                                                     }
                                                     else
                                                     {
                                                         // Other statuses --> Return error
                                                         if (handler) {
                                                             handler(nil, error);
                                                         }
                                                     }
                                                 }
                                                 else
                                                 {
                                                     // Parse failed --> Return error
                                                     if (handler) {
                                                         handler(nil, error);
                                                     }
                                                 }
                                             }
                                             else
                                             {
                                                 // Request failed --> Return error
                                                 if (handler) {
                                                     handler(nil, error);
                                                 }
                                             }
                                         }];
    [self.googleGeocoderTask resume];
}

- (NSArray *)buildSynchronousRequestFromURLString:(NSString *)urlString
{
    urlString = [urlString stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:[NSURL URLWithString:urlString]];
    
    NSURLResponse *response = nil;
    NSError *error = nil;
    NSData *data = [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:&error];
    if (!error && data)
    {
        NSError *parsingError = nil;
        NSDictionary *result = [NSJSONSerialization JSONObjectWithData:data
                                                               options:NSJSONReadingAllowFragments
                                                                 error:&parsingError];
        if (!parsingError && result)
        {
            NSString *status = [result objectForKey:@"status"];
            if ([status isEqualToString:@"OK"])
            {
                // Status OK --> Parse response results
                NSArray *locationDicts = [result objectForKey:@"results"];
                NSArray *finalResults = [self parseGeocodingResponseResults:locationDicts service:kLMGeocoderGoogleService];
                
                return finalResults;
            }
        }
    }
    
    return nil;
}


#pragma mark - PARSE RESULT DATA

- (NSArray *)parseGeocodingResponseResults:(NSArray *)responseResults service:(LMGeocoderService)service
{
    NSMutableArray *finalResults = [NSMutableArray new];
    
    for (id responseResult in responseResults) {
        LMAddress *address = [[LMAddress alloc] initWithLocationData:responseResult forServiceType:service];
        [finalResults addObject:address];
    }
    
    return finalResults;
}

@end

File: LMGeocoderSwift/LMGeocoderSwift/Classes/LMGeocoder.swift
//
//  LMGeocoder.swift
//  LMGeocoderSwift
//
//  Created by LMinh on 12/18/18.
//  Copyright © 2018 LMinh. All rights reserved.
//

import UIKit
import CoreLocation

/// LMGeocoder error codes, embedded in NSError.
public enum LMGeocoderErrorCode: Int {
    case InvalidCoordinate = 1, InvalidAddressString, Internal
}

/// LMGeocoder service API.
public enum LMGeocoderService {
    case GoogleService
    case AppleService
}

/// Handler that reports a geocoding response, or error.
public typealias LMGeocodeCallback = (_ results: Array<LMAddress>?, _ error: Error?) -> Void

/// Exposes a service for geocoding and reverse geocoding.
open class LMGeocoder: NSObject {
    
    // MARK: - PROPERTIES
    
    /// Indicating whether the receiver is in the middle of geocoding its value.
    private(set) var isGeocoding: Bool = false
    
    /// To set google API key.
    public var googleAPIKey: String? = nil
    
    /// Apple Geocoder.
    private lazy var appleGeocoder: CLGeocoder = CLGeocoder()
    
    /// Google Geocoder Task
    private lazy var googleGeocoderTask: URLSessionDataTask? = nil
    
    // MARK: SINGLETON
    
    public static let sharedInstance = LMGeocoder()
    
    // MARK: PUBLIC
    
    /// Submits a forward-geocoding request using the specified string.
    /// After initiating a forward-geocoding request, do not attempt to initiate another forward- or reverse-geocoding request.
    /// Geocoding requests are rate-limited for each app, so making too many requests in a
    /// short period of time may cause some of the requests to fail.
    /// When the maximum rate is exceeded, the geocoder passes an error object to your completion handler.
    ///
    /// - Parameters:
    ///   - address: The string describing the location you want to look up.
    ///   - service: The service API used to geocode.
    ///   - completionHandler: The callback to invoke with the geocode results. The callback will be invoked asynchronously from the main thread.
    open func geocode(address: String, service: LMGeocoderService, completionHandler: LMGeocodeCallback?) {
        
        self.isGeocoding = true
        
        // Check adress string
        guard address.count != 0 else {
            
            let error = NSError(domain: "LMGeocoderError",
                                code: LMGeocoderErrorCode.InvalidAddressString.rawValue,
                                userInfo: nil)
            
            self.isGeocoding = false
            if let completionHandler = completionHandler {
                completionHandler(nil, error)
            }
            return
        }
        
        // Check service
        if service == .GoogleService
        {
            // Geocode using Google service
            var urlString = "https://maps.googleapis.com/maps/api/geocode/json?sensor=true&address=\(address)"
            if let key = self.googleAPIKey {
                urlString = urlString + "&key=" + key
            }
            self.buildAsynchronousRequest(urlString: urlString) { (results, error) in
                
                self.isGeocoding = false
                if let completionHandler = completionHandler {
                    completionHandler(results, error)
                }
            }
        }
        else if service == .AppleService
        {
            // Geocode using Apple service
            self.appleGeocoder.geocodeAddressString(address) { (placemarks, error) in
                
                let results = self.parseGeocodingResponse(results: placemarks, service: .AppleService)
                
                self.isGeocoding = false
                if let completionHandler = completionHandler {
                    completionHandler(results, error)
                }
            }
        }
    }
    
    /// Submits a reverse-geocoding request for the specified coordinate.
    /// After initiating a reverse-geocoding request, do not attempt to initiate another reverse- or forward-geocoding request.
    /// Geocoding requests are rate-limited for each app, so making too many requests in a
    /// short period of time may cause some of the requests to fail.
    /// When the maximum rate is exceeded, the geocoder passes an error object to your completion handler.
    ///
    /// - Parameters:
    ///   - address: The coordinate to look up.
    ///   - service: The service API used to reverse geocode.
    ///   - completionHandler: The callback to invoke with the reverse geocode results.The callback will be invoked asynchronously from the main thread.
    open func reverseGeocode(coordinate: CLLocationCoordinate2D, service: LMGeocoderService, completionHandler: LMGeocodeCallback?) {
        
        self.isGeocoding = true
        
        // Check location coordinate
        guard CLLocationCoordinate2DIsValid(coordinate) else {
            
            let error = NSError(domain: "LMGeocoderError",
                                code: LMGeocoderErrorCode.InvalidCoordinate.rawValue,
                                userInfo: nil)
            
            self.isGeocoding = false
            if let completionHandler = completionHandler {
                completionHandler(nil, error)
            }
            return
        }
        
        // Check service
        if service == .GoogleService
        {
            // Reverse geocode using Google service
            var urlString = "https://maps.googleapis.com/maps/api/geocode/json?sensor=true&latlng=\(coordinate.latitude),\(coordinate.longitude)"
            if let key = self.googleAPIKey {
                urlString = urlString + "&key=" + key
            }
            self.buildAsynchronousRequest(urlString: urlString) { (results, error) in
                
                self.isGeocoding = false
                if let completionHandler = completionHandler {
                    completionHandler(results, error)
                }
            }
        }
        else if service == .AppleService
        {
            // Reverse geocode using Apple service
            let location = CLLocation(latitude: coordinate.latitude, longitude: coordinate.longitude)
            self.appleGeocoder.reverseGeocodeLocation(location) { (placemarks, error) in
                
                let results = self.parseGeocodingResponse(results: placemarks, service: .AppleService)
                
                self.isGeocoding = false
                if let completionHandler = completionHandler {
                    completionHandler(results, error)
                }
            }
        }
    }
    
    /// Cancels a pending geocoding request.
    open func cancelGeocode() {
        self.appleGeocoder.cancelGeocode()
        self.googleGeocoderTask?.cancel()
    }
    
    // MARK: INTERNAL
    
    /// Build asynchronous request
    func buildAsynchronousRequest(urlString: String, completionHandler: LMGeocodeCallback?) {
        
        let urlString = urlString.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)!
        let url = URL(string: urlString)!
        
        self.googleGeocoderTask = URLSession.shared.dataTask(with: url, completionHandler: { (data, response, error) in
            
            if let data = data, error == nil {
                
                do {
                    let result = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as! Dictionary<String, Any>
                    
                    // Check status value
                    if let status = result["status"] as? String, status == "OK" {
                        // Status OK --> Parse response results
                        let locationDicts = result["results"] as? Array<AnyObject>
                        let finalResults = self.parseGeocodingResponse(results: locationDicts, service: .GoogleService)
                        if let completionHandler = completionHandler {
                            completionHandler(finalResults, nil)
                        }
                    }
                    else {
                        // Other status --> Return error
                        if let completionHandler = completionHandler {
                            completionHandler(nil, error)
                        }
                    }
                }
                catch {
                    // Parse failed --> Return error
                    if let completionHandler = completionHandler {
                        completionHandler(nil, error)
                    }
                }
            }
            else {
                // Request failed --> Return error
                if let completionHandler = completionHandler {
                    completionHandler(nil, error)
                }
            }
        })
        self.googleGeocoderTask?.resume()
    }
    
    /// Parse geocoding response
    func parseGeocodingResponse(results: Array<AnyObject>?, service: LMGeocoderService) -> Array<LMAddress>? {
        
        guard let results = results else { return nil }
        
        var finalResults = Array<LMAddress>()
        for respondObject in results {
            let address = LMAddress(locationData: respondObject, serviceType: service)
            finalResults.append(address)
        }
        return finalResults
    }
}
